BUG: integrate: make `select_initial_step` aware of integration interval

valentineap opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes gh-17341 
Closes gh-8848
Closes gh-9198

#### What does this implement/fix?
<!--Please explain your changes.-->
#17341 reports that `solve_ivp(func, (t0, t_bound), f0)` would evaluate `func()` outside the range `(t0,  t_bound)` if `|t_bound-t0|` is small. 

This issue was traced to `select_initial_step`, which was not aware of the far limit, `t_bound`, and evaluates `func(t0+h0)` for some internally-determined value of `h0`. 

Fix:
- add optional argument `t_bound` to `select_initial_step` and ensure that that `h0<=|t_bound-t0|`
- modify all solvers within `_ivp` that use `select_initial_step` to take advantage of this argument

#### Additional information
<!--Any additional information you think is important.-->

If `t0 == t_bound`, `select_initial_step` will now return 0. This is a logical result but has potential for divide-by-zero issues.

valentineap commented:
Hi @Patol75, in my view it is a clear bug, as there is no requirement that the function `func` is well-defined (mathematically or computationally) outside the domain of integration. The integrator should never attempt to evaluate `func` outside the domain, as doing so may result in errors or other unexpected behaviour.

Thus the need for a more complex solution than you propose -- the point is to avoid ever calling `func` for these values, not just to avoid using them.

Patol75 commented:
That is a fair point indeed. The current implementation would require users to handle function calls outside of the expected domain in the case of exotic functions, which does not seem ideal. Nonetheless, I am still wondering if this is a known issue with regards to the implemented scheme, in which case applying a patch seems reasonable, or if there is a mistake in the implementation of the scheme.

@mdhaber Do you mind sharing your opinion on these changes?

valentineap commented on scipy/integrate/_ivp/common.py:
```diff
@@ -108,7 +111,12 @@ def select_initial_step(fun, t0, y0, f0, direction, order, rtol, atol):
+        if h0 > (t_bound-t0)/direction: 
+            h0 = (t_bound-t0)/direction
+        # Case where t0 = t_bound
+        if h0 == 0: return 0.
```
In principle, yes -- but I think that would then necessitate refactoring some of the integration routines to handle `tbound == t0` as a special case; at present they call `select_initial_step` and do various other things before determining that they should return 0.

valentineap replied:
I think the current version works ok with the solvers as they stand -- the existing test-suite includes the case `tbound==t0`.

Patol75 commented:
I have just had another look at this PR. I believe there are two main issues highlighted here:
- `select_initial_step` can return values that exceed the integration span
- Null integration spans are left unchecked

For the second item, it is true that `integrate` solvers can handle a null span and return without errors claiming that they have successfully reached the end of the integration interval. I am not quite sure if this is a meaningful way to deal with such a situation. The immediate alternative I am seeing is to add a check in [`check_arguments`](https://github.com/scipy/scipy/blob/cfc8958d1933de9a848c78703cea814e6ff0ce90/scipy/integrate/_ivp/base.py#L4) to interrupt execution if there is basically nothing to be done.

For the first item, `select_initial_step` is only called by RK-family solvers, Radau, and BDF — LSODA has its own first-step selection algorithm, which does not suffer from the issue highlighted here, as reported by @laurent90git. The five mentioned solvers all execute the following code:
```
        if first_step is None:
            self.h_abs = select_initial_step(
                self.fun, self.t, self.y, self.f, self.direction,
                self.error_estimator_order, self.rtol, self.atol)
        else:
            self.h_abs = validate_first_step(first_step, t0, t_bound)
```
We could change the above to
```
        if first_step is None:
            first_step = select_initial_step(
                self.fun, self.t, self.y, self.f, self.direction,
                self.error_estimator_order, self.rtol, self.atol)
        self.h_abs = validate_first_step(first_step, t0, t_bound)
```
Within `validate_first_step`, we have
```
    if first_step <= 0:
        raise ValueError("`first_step` must be positive.")
    if first_step > np.abs(t_bound - t0):
        raise ValueError("`first_step` exceeds bounds.")
    return first_step
```
This could change to
```
    if first_step <= 0:
        raise ValueError("`first_step` must be positive.")
    return min(first_step, np.abs(t_bound - t0))
```

What would you think of such a change? If you do not believe it is a good idea, then any objections to the changes proposed by @valentineap?

valentineap commented:
As a matter of general principle: it seems reasonable to handle the case where `t_0 == t_bound` at the earliest possible opportunity -- it is axiomatic that the result should be zero and no work needs to be done to discover that.

As for @Patol75's suggested fix to the `validate_first_step` function: this is a straightforward fix to the problem I identified. However, the downside is that if the user has chosen to explicitly specify their own value for `first_step` but has provided an inappropriate value, this will silently be overridden. I am not sure this is desirable -- if a user seeks control over this option they would probably prefer to be alerted that their chosen value is unsuitable.

mdhaber commented on scipy/integrate/_ivp/common.py:
```diff
@@ -108,7 +111,12 @@ def select_initial_step(fun, t0, y0, f0, direction, order, rtol, atol):
-
+    if t_bound is not None:
+        # Check t0+h0*direction doesn't take us beyond t_bound
+        if h0 > (t_bound-t0)/direction: 
+            h0 = (t_bound-t0)/direction
```
Nit: the logic `h = min(h, (t_bound-t0)/direction)` is used below. Consider using the same here idea here.

valentineap replied:
Fixed.

mdhaber commented on scipy/integrate/_ivp/common.py:
```diff
@@ -63,7 +63,7 @@ def norm(x):
 
 
-def select_initial_step(fun, t0, y0, f0, direction, order, rtol, atol):
+def select_initial_step(fun, t0, y0, f0, direction, order, rtol, atol, t_bound=None):
```
Despite not having an underscore, `select_initial_step` is [not public](https://docs.scipy.org/doc/scipy/reference/), so if it doesn't break any uses of the function within SciPy, it is arguably OK to change this without a default. Are all the uses of `select_initial_step` inside SciPy addressed here?

mdhaber reviewed (commented):
> Do you mind sharing your opinion on these changes?

Oops, I missed the ping.

Does this address gh-9198 and/or gh-8848? If so, I'd support such a change. 

That said, I haven't taken a close look at how this works, and I'd need to ask some basic questions to facilitate review. LMK if you'd like me to take a closer look.

Patol75 commented:
> > Do you mind sharing your opinion on these changes?
> 
> Oops, I missed the ping.
> 
> Does this address [gh-9198](https://github.com/scipy/scipy/issues/9198) and/or [gh-8848](https://github.com/scipy/scipy/issues/8848)? If so, I'd support such a change.
> 
> That said, I haven't taken a close look at how this works, and I'd need to ask some basic questions to facilitate review. LMK if you'd like me to take a closer look.

Yes, the changes here would prevent any function evaluation outside the integration range specified by the user. It is relatively straightforward to do so, either in the way the PR proposes or the way my message above suggests. But, as I stressed in my first reply here, it is not necessarily obvious for such behaviour to be a bug — consider, for example, @WarrenWeckesser [reply](https://github.com/scipy/scipy/issues/9198#issuecomment-420723549) in gh-9198.

mdhaber commented:
I'll limit my reply to the original concern: "`solve_ivp(func, (t0, t_bound), f0)` would **evaluate** `func()` outside the range `(t0,  t_bound)`".  You mentioned two separate issues in https://github.com/scipy/scipy/pull/17348#issuecomment-1451111400, but I haven't considered them.

> But, as I stressed in my first reply here, it is not necessarily obvious for such behaviour to be a bug

@drhagen was involved in creating the solvers and addressed that in https://github.com/scipy/scipy/issues/9198#issuecomment-422760932, stating that it was a bug. After more carefully reading the code and related issues, I think it is something to address, whether it is called a bug or not, but I lean toward "bug" because it seems to have been inadvertent and does not seem to be a useful "feature".

Looking at the original discussion (e.g. [here](https://github.com/scipy/scipy/pull/6326#issuecomment-241244680)), `t_bound` was initially known as `t_crit`, and it was supposed to be the time beyond which the integrand would _not be evaluated because the integrate might not be well defined beyond this time. I agree that this intent is not 100% explicit in the description "Boundary time — the integration won’t continue beyond it.", but it seems to be clarified in the code. There are several places where it is used to prevent function evaluation beyond the specified time.

https://github.com/scipy/scipy/blob/c1ed5ece8ffbf05356a22a8106affcd11bd3aee0/scipy/integrate/_ivp/rk.py#L138-L139

In any case, SUNDIALS apparently has an option to turn this behavior on and off. We don't, so something should change.

If there is a demonstrable benefit to evaluating the integrand beyond `t_bound`, let's reconsider, but right now, it seems like the intent was to prevent evaluation of the integrand beyond `t_bound` (not just to prevent steps after it is exceeded). This has come up in three issues, and this PR sounds like a reasonable solution to the problem, so I'd support it. In addition, I would prevent function evaluations beyond `t0 + max_step`. 

(One could argue that _only_ preventing function evaluations beyond `max_step` from the initial/current time would be sufficient. I agree that it would be enough to allow the user to prevent `select_initial_step` from evaluating the integrand beyond `t_bound`, but I still think the default should be to prevent this without requiring `max_step`.)

valentineap commented:
I completely agree with @mdhaber. I think it is important to highlight that this issue (at least #17348; I haven't fully investigated the various other linked issues) is solely associated with the `select_initial_step` routine. If a user specifies their own step this gets passed into `validate_initial_step` which raises an error if this is larger than the integration domain.

mdhaber commented:
@valentineap Great. In that case, please consider:
- adjusting this PR to ensure that `max_step` is respected in `select_initial_step` and demonstrating this with a unit test
- investigating whether you believe this closes gh-9198 and/or gh-8848. If so, please add any associated tests. (For instance, gh-8848 has some simple code that could be adapted to become a unit test. You might check that the code runs without emitting a warning and that the final values of `y` are close to those produced by `odeint`.)

valentineap commented:
Closes gh-8848.
Closes gh-9198.

valentineap commented:
@mdhaber OK, finally getting around to dealing with this..

This PR modifies `scipy.integrate._ivp.common.select_initial_step()` to ensure that:
- The inital step is no larger than the requested integration interval, and
- The initial step is no larger than the specified maximum integration step.

It will close #8848, #9198, and #17341. Example code from each of these issues is now included in the test suite: `test_ivp.test_tbound_larger_interval()`, `test_ivp.test_tbound_oscillator()` and `test_ivp.test_tbound_small_interval()` respectively. Additionally `test_ivp.test_inital_maxstep()` verifies that `max_step` is honoured.

Enforcing `max_step` within `select_initial_step` broke one existing test: `test_ivp.test_max_step()`. This includes a test that integration fails if `max_step` is too small. It assumed failure on the first step. With the change, the first step succeeds and failure occurs on the second step. It seems to me that this is still acceptable from the point of view of 'passing' the test, and I have modified it accordingly.

I think this is ready to be considered for merge into main.

tylerjereddy commented:
> I think this is ready to be considered for merge into main.

I'm not disagreeing with this, but I'm going to bump the milestone because I don't see enough recent review activity to give me the confidence to hit the green button.

If this does get reviewed/merged before we branch for `1.14.0`, we can set the milestone back.

valentineap commented:
@tylerjereddy Understood. But perhaps you could suggest how to stimulate some review activity? This seems to have dropped off everyone's radar. I think this is a fixed bug and it would be good to get it out there.

nmayorov commented on scipy/integrate/_ivp/common.py:
```diff
@@ -108,7 +114,10 @@ def select_initial_step(fun, t0, y0, f0, direction, order, rtol, atol):
+    # Check t0+h0*direction doesn't take us beyond t_bound
+    h0 = min(h0, (t_bound-t0)/direction)
+    # Case where t0 = t_bound
+    if h0 == 0: return 0.
```
Perhaps move this case right after `if y0.size == 0: ...` as a very degenerate.

Also probably better not to use one-liner ifs as they are not often used in the code base.

nmayorov commented on scipy/integrate/_ivp/common.py:
```diff
@@ -108,7 +114,10 @@ def select_initial_step(fun, t0, y0, f0, direction, order, rtol, atol):
         h0 = 0.01 * d0 / d1
-
+    # Check t0+h0*direction doesn't take us beyond t_bound
+    h0 = min(h0, (t_bound-t0)/direction)
```
I'd use `abs` here or multiplying by direction.

nmayorov commented on scipy/integrate/_ivp/common.py:
```diff
@@ -118,7 +127,10 @@ def select_initial_step(fun, t0, y0, f0, direction, order, rtol, atol):
         h1 = (0.01 / max(d1, d2)) ** (1 / (order + 1))
 
-    return min(100 * h0, h1)
+    h = min(100 * h0, h1)
```
Python supports `min` with many arguments, I'd condense everything in one line. I think it will be quite readable and easy to understand. 

Also might denote `abs(t_bound - to)` as some parameter as it's used 2 times.

nmayorov commented on scipy/integrate/_ivp/tests/test_ivp.py:
```diff
@@ -1091,3 +1091,79 @@ def test_initial_state_finiteness(f0_fill):
+    # f[y(t)] = 2y(t) on t in [0,SMALL]
+    #           undefined otherwise 
+    def f(t,y):
+        if t>SMALL:raise ValueError("Function was evaluated outside interval")
```
Don't want to be overly picky, but the code looks a bit untidy.

Please use standard `if` formatting, add spaces around all operators, use `0.0` to denote floats.

valentineap replied:
Hopefully now better; most of this code came from the various bug reports.

nmayorov commented on scipy/integrate/_ivp/tests/test_ivp.py:
```diff
@@ -1091,3 +1091,79 @@ def test_initial_state_finiteness(f0_fill):
+    res = solve_ivp(f, (0,SMALL), np.array([1]),method=method)
+    assert_(res.success)
+
+@pytest.mark.parametrize('method', ['RK23', 'RK45', 'DOP853', 'Radau', 'BDF'])
```
Probably 2 spaces between functions in the module.

nmayorov reviewed (changes requested):
I believe conceptually it is a correct change:

- It make that integration bounds are respected everywhere
- Reducing the step size to fit bounds or requested max_step is logical and in principle shouldn't affect the solvers in a negative way

Regarding the case `t == t_bound`, could you please add a test that it doesn't crash for this case?

j-bowhay commented on scipy/integrate/_ivp/tests/test_ivp.py:
```diff
@@ -1085,9 +1085,116 @@ def fun_with_arg(t, y, a):
+                         dense_output = True, 
+                         max_step = t1 - t0)
+    result = run_sim2(1000, 100, 100)
+    assert_(result.success)
```
Here and in the other new tests added we now favour plain asserts
```suggestion
    assert result.success
```

valentineap commented:
@nmayorov Thanks for the comments -- I believe I have now addressed all of these. Let me know if you think I've missed anything.

@j-bowhay I have updated the `assert` statements in the tests that are affected by this PR; I haven't looked at any of the pre-existing tests.

nmayorov commented on scipy/integrate/_ivp/common.py:
```diff
@@ -107,6 +107,10 @@ def select_initial_step(fun, t0, y0, t_bound, max_step, f0, direction, order, rt
         return np.inf
 
+    interval_length = abs(t_bound - t0)
+    if interval_length == 0.:
```
Sorry, can we use 0.0 to denote floats? Not sure if it's a hard requirement by any Python linter, but I find it much nicer to read.

nmayorov commented:
Looks much cleaner now.

What about a test for `t0 == t_bound`, wouldn't it be too hard to implement it?

valentineap commented:
@nmayorov I had added `test_zero_interval` which checks the case `t0 == t_bound`. Perhaps I'm misunderstanding what you wanted?

I've fixed the `0. -> 0.0` issue.

nmayorov commented:
> @nmayorov I had added `test_zero_interval` which checks the case `t0 == t_bound`. Perhaps I'm misunderstanding what you wanted?
> 
> I've fixed the `0. -> 0.0` issue.

Just didn't spot it quickly, excellent.

valentineap commented:
Great -- so unless there's anything I've missed, I think that's all of @nmayorov's comments addressed.

I don't think the 'failing' check is anything to do with the changes made in this PR.

@tylerjereddy Perhaps this is now suitable for merge?

nmayorov commented on scipy/integrate/_ivp/bdf.py:
```diff
@@ -204,9 +204,11 @@ def __init__(self, fun, t0, y0, t_bound, max_step=np.inf,
                                              self.direction, 1,
-                                             self.rtol, self.atol)
+                                             self.rtol, self.atol 
+                                             )
```
No need to move ) here.

nmayorov commented on scipy/integrate/_ivp/common.py:
```diff
@@ -88,6 +93,7 @@ def select_initial_step(fun, t0, y0, f0, direction, order, rtol, atol):
     atol : float
         Desired absolute tolerance.
 
+
```
No need to add this line.

nmayorov commented on scipy/integrate/_ivp/common.py:
```diff
@@ -118,7 +130,8 @@ def select_initial_step(fun, t0, y0, f0, direction, order, rtol, atol):
         h1 = (0.01 / max(d1, d2)) ** (1 / (order + 1))
 
-    return min(100 * h0, h1)
+    h = min(100 * h0, h1, interval_length, max_step)
```
Perhaps just `return` as was before?

nmayorov reviewed (approved):
Some very minor comments.

j-bowhay commented on scipy/integrate/_ivp/tests/test_ivp.py:
```diff
@@ -7,7 +7,7 @@
 from scipy.integrate import solve_ivp, RK23, RK45, DOP853, Radau, BDF, LSODA
 from scipy.integrate import OdeSolution
-from scipy.integrate._ivp.common import num_jac
+from scipy.integrate._ivp.common import num_jac,select_initial_step
```
```suggestion
from scipy.integrate._ivp.common import num_jac, select_initial_step
```

j-bowhay commented on scipy/integrate/_ivp/tests/test_ivp.py:
```diff
@@ -555,7 +555,7 @@ def test_max_step():
                                 rtol=rtol, atol=atol, max_step=1e-20)
                 message = solver.step()
-
+                message = solver.step() # First step succeeds but second step fails.
```
```suggestion
                message = solver.step()  # First step succeeds but second step fails.
```

j-bowhay commented on scipy/integrate/_ivp/tests/test_ivp.py:
```diff
@@ -1085,9 +1085,117 @@ def fun_with_arg(t, y, a):
+    # f[y(t)] = 2y(t)
+    def f(t, y):
+        return 2 * y
+    res = solve_ivp(f, (0.0, 0.0), np.array([1.0]), method = method)
```
```suggestion
    res = solve_ivp(f, (0.0, 0.0), np.array([1.0]), method=method)
```

j-bowhay commented on scipy/integrate/_ivp/tests/test_ivp.py:
```diff
@@ -1085,9 +1085,117 @@ def fun_with_arg(t, y, a):
+        return 2 * y
+    res = solve_ivp(f, (0.0, 0.0), np.array([1.0]), method = method)
+    assert res.success
+    assert_allclose(res.y[0,-1], 1.0)
```
```suggestion
    assert_allclose(res.y[0, -1], 1.0)
```

valentineap commented:
Thanks @nmayorov @j-bowhay -- I've incorporated your formatting fixes.

Note that this PR should close gh-8848 and gh-9198 in addition to gh-17341 for which it is currently marked.

j-bowhay commented:
@valentineap can you edit the description linking those issues too please

valentineap commented:
@j-bowhay Done.

nmayorov commented:
To me everything looks good.

nmayorov commented:
I'll go ahead and merge it.

I believe it's a great fix closing a lot of issues. True, there is always a risk of breaking something in numerical algorithms, but this change looks rather safe and makes the `t_bound` and `max_step` arguments properly defined.

Thank you @valentineap for the work and insisting on finishing it.

@tylerjereddy  can be included in 0.14.0 then.
